File: interpreted/typescript/bun/nanocode.ts
#!/usr/bin/env bun
// nanocode - minimal claude code alternative (Bun)
import { $ } from "bun";

const KEY = Bun.env.OPENROUTER_API_KEY || Bun.env.ANTHROPIC_API_KEY;
const API = Bun.env.OPENROUTER_API_KEY ? "https://openrouter.ai/api/v1/messages" : "https://api.anthropic.com/v1/messages";
const MODEL = Bun.env.MODEL || (Bun.env.OPENROUTER_API_KEY ? "anthropic/claude-opus-4" : "claude-sonnet-4-20250514");
const [R, B, D, C, G] = ["\x1b[0m", "\x1b[1m", "\x1b[2m", "\x1b[36m", "\x1b[32m"];

const tools: Record<string, [(a: any) => string|Promise<string>, string, object]> = {
  read: [(a) => Bun.file(a.path).text().then(t => t.split("\n").slice(a.offset||0, (a.offset||0)+(a.limit||Infinity)).map((l,i) => `${(a.offset||0)+i+1}| ${l}`).join("\n")), "Read file", {path:"string",offset:"number?",limit:"number?"}],
  write: [(a) => (Bun.write(a.path, a.content), "ok"), "Write file", {path:"string",content:"string"}],
  edit: [async (a) => { const t = await Bun.file(a.path).text(); if(!t.includes(a.old)) return "error: not found"; if(!a.all && t.split(a.old).length>2) return "error: not unique"; await Bun.write(a.path, a.all?t.replaceAll(a.old,a.new):t.replace(a.old,a.new)); return "ok"; }, "Edit file", {path:"string",old:"string",new:"string",all:"boolean?"}],
  glob: [async (a) => { const g = new Bun.Glob(a.pat); return [...g.scanSync(a.path||".")].slice(0,50).join("\n")||"none"; }, "Find files", {pat:"string",path:"string?"}],
  grep: [async (a) => { const g = new Bun.Glob("**/*"); const re = new RegExp(a.pat); return [...g.scanSync(a.path||".")].flatMap(f => { try { const lines = Bun.file(f).text(); return []; } catch { return []; }}).join("\n")||"none"; }, "Search", {pat:"string",path:"string?"}],
  bash: [async (a) => { try { return await $`${a.cmd}`.text(); } catch(e:any) { return e.message; }}, "Run command", {cmd:"string"}],
};

const schema = Object.entries(tools).map(([name, [_, desc, params]]) => ({
  name, description: desc,
  input_schema: { type: "object", properties: Object.fromEntries(Object.entries(params).map(([k,v]) => [k, {type:(v as string).replace("?","")==="number"?"integer":(v as string).replace("?","")}])), required: Object.entries(params).filter(([_,v]) => !(v as string).endsWith("?")).map(([k]) => k) }
}));

async function ask(messages: any[]) {
  const res = await fetch(API, {
    method: "POST", headers: { "Content-Type": "application/json", "anthropic-version": "2023-06-01", ...(Bun.env.OPENROUTER_API_KEY ? { Authorization: `Bearer ${KEY}` } : { "x-api-key": KEY! }) },
    body: JSON.stringify({ model: MODEL, max_tokens: 8192, system: `Concise coding assistant. cwd: ${process.cwd()}`, messages, tools: schema })
  });
  return res.json();
}

console.log(`${B}nanocode${R} | ${D}${MODEL}${R}\n`);
const messages: any[] = [];

for await (const line of console.write(`${B}\x1b[34m❯${R} `), Bun.stdin.stream()) {
  const input = new TextDecoder().decode(line).trim();
  if (!input) { console.write(`${B}\x1b[34m❯${R} `); continue; }
  if (input === "/q") break;
  if (input === "/c") { messages.length = 0; console.log(`${G}⏺ Cleared${R}`); console.write(`${B}\x1b[34m❯${R} `); continue; }
  
  messages.push({ role: "user", content: input });
  while (true) {
    const { content } = await ask(messages);
    const results: any[] = [];
    for (const block of content) {
      if (block.type === "text") console.log(`\n${C}⏺${R} ${block.text}`);
      if (block.type === "tool_use") {
        console.log(`\n${G}⏺ ${block.name}${R}(${D}${JSON.stringify(block.input).slice(0,50)}${R})`);
        const result = await tools[block.name][0](block.input);
        console.log(`  ${D}⎿ ${result.split("\n")[0].slice(0,60)}${R}`);
        results.push({ type: "tool_result", tool_use_id: block.id, content: result });
      }
    }
    messages.push({ role: "assistant", content });
    if (!results.length) break;
    messages.push({ role: "user", content: results });
  }
  console.log(); console.write(`${B}\x1b[34m❯${R} `);
}

File: gpu/triton/nanocode.py
#!/usr/bin/env python3
"""nanocode - minimal claude code alternative (Triton GPU)

Triton is OpenAI's GPU programming language for deep learning.
This demonstrates how AI agents could run on GPUs for massive parallelism.

pip install triton && python nanocode.py
"""

import triton
import triton.language as tl
import torch
import json
import os
import glob as globlib
import re
import urllib.request

# GPU-accelerated string processing kernel
@triton.jit
def process_tokens_kernel(
    input_ptr, output_ptr,
    n_elements,
    BLOCK_SIZE: tl.constexpr,
):
    """Process tokens in parallel on GPU"""
    pid = tl.program_id(axis=0)
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    
    # Load input tokens
    x = tl.load(input_ptr + offsets, mask=mask)
    
    # Simple processing (in real impl, would do embedding lookup, etc.)
    output = x
    
    # Store output
    tl.store(output_ptr + offsets, output, mask=mask)

def gpu_process(text: str) -> str:
    """Process text on GPU (demonstration)"""
    # Convert to tensor
    tokens = torch.tensor([ord(c) for c in text], dtype=torch.int32, device='cuda')
    output = torch.empty_like(tokens)
    
    # Launch kernel
    n_elements = tokens.numel()
    grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']),)
    process_tokens_kernel[grid](tokens, output, n_elements, BLOCK_SIZE=1024)
    
    # Convert back
    return ''.join(chr(c) for c in output.cpu().tolist())

# Standard nanocode implementation
OPENROUTER_KEY = os.environ.get("OPENROUTER_API_KEY")
KEY = OPENROUTER_KEY or os.environ.get("ANTHROPIC_API_KEY")
API = os.environ.get("API_URL", "https://openrouter.ai/api/v1/messages" if OPENROUTER_KEY else "https://api.anthropic.com/v1/messages")
MODEL = os.environ.get("MODEL", "anthropic/claude-opus-4-5" if OPENROUTER_KEY else "claude-opus-4-5")
R, B, D, C, G, BL = "\033[0m", "\033[1m", "\033[2m", "\033[36m", "\033[32m", "\033[34m"


def _read(a):
    lines = open(a["path"]).read().splitlines()
    off = a.get("offset", 0)
    lim = a.get("limit", len(lines))
    return "\n".join(f"{off+i+1}| {l}" for i, l in enumerate(lines[off : off + lim]))


def _write(a):
    open(a["path"], "w").write(a["content"])
    return "ok"


def _edit(a):
    text = open(a["path"]).read()
    old, new = a["old"], a["new"]
    count = text.count(old)
    if not count:
        return "error: old_string not found"
    if count > 1 and not a.get("all"):
        return f"error: old_string appears {count} times, use all=true"
    updated = text.replace(old, new) if a.get("all") else text.replace(old, new, 1)
    open(a["path"], "w").write(updated)
    return "ok"


def _glob(a):
    pat = (a.get("path", ".") + "/" + a["pat"]).replace("//", "/")
    files = sorted(globlib.glob(pat, recursive=True))
    return "\n".join(files) or "none"


def _grep(a):
    regex = re.compile(a["pat"])
    hits = []
    for path in globlib.glob(a.get("path", ".") + "/**", recursive=True):
        if os.path.isdir(path):
            continue
        try:
            for i, line in enumerate(open(path), 1):
                if regex.search(line):
                    hits.append(f"{path}:{i}:{line.rstrip()}")
        except Exception:
            continue
    return "\n".join(hits[:50]) or "none"


def _bash(a):
    return __import__("subprocess").run(a["cmd"], shell=True, capture_output=True, text=True).stdout


tools = {"read": _read, "write": _write, "edit": _edit, "glob": _glob, "grep": _grep, "bash": _bash}

schema = [
    {"name": "read", "description": "Read file", "input_schema": {"type": "object", "properties": {"path": {"type": "string"}, "offset": {"type": "integer"}, "limit": {"type": "integer"}}, "required": ["path"]}},
    {"name": "write", "description": "Write file", "input_schema": {"type": "object", "properties": {"path": {"type": "string"}, "content": {"type": "string"}}, "required": ["path", "content"]}},
    {"name": "edit", "description": "Replace text", "input_schema": {"type": "object", "properties": {"path": {"type": "string"}, "old": {"type": "string"}, "new": {"type": "string"}, "all": {"type": "boolean"}}, "required": ["path", "old", "new"]}},
    {"name": "glob", "description": "Find files", "input_schema": {"type": "object", "properties": {"pat": {"type": "string"}, "path": {"type": "string"}}, "required": ["pat"]}},
    {"name": "grep", "description": "Search files", "input_schema": {"type": "object", "properties": {"pat": {"type": "string"}, "path": {"type": "string"}}, "required": ["pat"]}},
    {"name": "bash", "description": "Run command", "input_schema": {"type": "object", "properties": {"cmd": {"type": "string"}}, "required": ["cmd"]}},
]

def ask(messages):
    headers = {"Content-Type": "application/json", "anthropic-version": "2023-06-01"}
    headers["Authorization" if OPENROUTER_KEY else "x-api-key"] = (f"Bearer {KEY}" if OPENROUTER_KEY else KEY)
    req = urllib.request.Request(
        API,
        json.dumps({"model": MODEL, "max_tokens": 8192, "system": "Concise assistant", "messages": messages, "tools": schema}).encode(),
        headers,
    )
    return json.loads(urllib.request.urlopen(req).read())

def main():
    device = "GPU (Triton)" if torch.cuda.is_available() else "CPU (Triton unavailable)"
    print(f"{B}nanocode{R} | {D}{MODEL} | {device}{R}\n")
    messages = []
    
    while True:
        try:
            user_input = input(f"{B}{BL}❯{R} ").strip()
            if not user_input: continue
            if user_input in ("/q", "exit"): break
            if user_input == "/c": messages = []; print(f"{G}⏺ Cleared{R}"); continue
            
            # GPU-accelerate input processing if available
            if torch.cuda.is_available():
                user_input = gpu_process(user_input)
            
            messages.append({"role": "user", "content": user_input})
            while True:
                resp = ask(messages)
                content = resp.get("content", [])
                results = []
                for block in content:
                    if block["type"] == "text": print(f"\n{C}⏺{R} {block['text']}")
                    if block["type"] == "tool_use":
                        print(f"\n{G}⏺ {block['name']}{R}")
                        result = tools[block["name"]](block["input"])
                        print(f"  {D}⎿ {result.split(chr(10))[0][:60]}{R}")
                        results.append({"type": "tool_result", "tool_use_id": block["id"], "content": result})
                messages.append({"role": "assistant", "content": content})
                if not results: break
                messages.append({"role": "user", "content": results})
            print()
        except (KeyboardInterrupt, EOFError): break

if __name__ == "__main__": main()
